# 5.1 Built-in Functions

Built-in functions provides specific operations that come out of the box. Guess what, We have already used built-in function, that's none other than our beloved `print` function, I guess everyone's first try of built-in function would be `print` function😎. 

Thanks to Python 🐍 for giving a huge number of built-in functions. Builtin functions need not be imported explicitly.Let's get the list of built-in functions available with Python

In [ ]:
import builtins
from types import BuiltinFunctionType

all_builtin_function_names_objects = [
    name
    for name, obj in builtins.__dict__.items()
    if isinstance(obj, BuiltinFunctionType)
]
print(*all_builtin_function_names_objects, sep="\n")

The question that hits 👊 our brain now is "How did we get the above output?". Well, that's simple, we imported the builtins and generated a list by looping over the attributes for each object name `name` and object `obj` and adding the name of the object to the list `all_builtin_function_names_objects` if the object is an instance of `BuiltinFunctionType`.

Let's get a sip of each of the built-in functions

## \_\_build_class\_\_

`__build_class__` is an internal helper function to create a class. Whenever a object of a class is created, `__build__class__` method gets called implicitly.

## \_\_import\_\_

## abs

`abs` function returns the absolute value of int, float and complex numbers. Absolute value doesn't have sign, it is always a non-negative value.

In [23]:
# abs on positive integer.
print(abs(5))

5


In [24]:
# abs on negative integer.
print(abs(-5))

5


In [25]:
# abs on float.
print(abs(-2.34))

2.34


In [26]:
# abs on complex.
print(abs(2-4j))

4.47213595499958


💡 Just in case, if we have want to know how we got the `abs(2-4j)` as 4.47213595499958. Absolute value of complex number x+yi is calculated as 

```math
abs(x+yi) = sqrt(x^2 + y^2)
```

coming to our example:
```math
abs(2-4j) = sqrt(2^2 + 4^2)
= sqrt(4 + 16)
= sqrt(20)
= 4.47213595499958

```

## all

`all` function accepts iterables like list, tuple, etc.. and if boolean equivalent of each object in the iterable is `True` then `all` function returns `True`. If atleast one object's boolean equivalent is `False`, then `all` function returns `False`.

In [28]:
# all on [0].
print(all([0]))

False


boolean equivalent of 0 is `False`

In [29]:
# all on [0,1]
print(all([0, 1]))

False


In [30]:
# all on [1,2,3]
print(all([1,2,3]))

True


In [33]:
# all on [1,["abcd"], "Hello", "Python", "🐍"]
print(all([1,["abcd"], "Hello", "Python", "🐍"]))

True


In [1]:
# all on []
print(all([]))

True


`all` function on empty iterable returns `True`.

## any

`any` function is similar to `all` function, just that `any` function returns `True` if atleast one object's boolean equivalent is `True`, whereas in `all`, every obejct's boolean equivalent should be `True`.

In [1]:
# any on [0].
print(any([0]))

False


boolean equivalent of 0 is `False`

In [2]:
# any on [0,1]
print(any([0, 1]))

True


In [3]:
# any on [1,2,3]
print(any([1,2,3]))

True


In [4]:
# any on [1,["abcd"], "Hello", "Python", "🐍"]
print(any([1,["abcd"], "Hello", "Python", "🐍"]))

True


In [5]:
# any on []
print(any([]))

False


💡 Note that `any` on empty list returned `False`, but `all` on empty list returned `True`.

## ascii

In [10]:
repr("a\t")

"'a\\t'"

## bin

## breakpoint

## callable

## chr

## compile

## delattr

## dir

## divmod

## eval

## exec

## format

## getattr

## globals

## hasattr

## hash

## hex

## id

## isinstance

## issubclass

## iter

## len

## locals

## max

## min

## next

## oct

## ord

## pow

## print

## repr

## round

## setattr

## sorted

## sum

The function name `sum` itself 

## vars

## open